app/lib/shared/search_service.dart

```diff
@@ -0,0 +1,156 @@
+  final String description;
+  final String lastUpdated;
+  final String readme;
+
+  final List<String> detectedTypes;
+  final double popularity;
+
+  PackageDocument({
+    this.url,
+    this.package,
+    this.version,
+    this.devVersion,
+    this.description,
+    this.lastUpdated,
+    this.readme,
+    this.detectedTypes,
+    this.popularity,
+  });
+
+  factory PackageDocument.parseJson(Map json) => new PackageDocument(
+        url: json['url'],
+        package: json['package'],
+        version: json['version'],
+        devVersion: json['devVersion'],
+        description: json['description'],
+        lastUpdated: json['lastUpdated'],
+        readme: json['readme'],
+        detectedTypes: json['detectedTypes'],
+        popularity: json['popularity'],
+      );
+
+  Map toJson() => {
+        'url': url,
+        'package': package,
+        'version': version,
+        'devVersion': devVersion,
+        'description': description,
+        'lastUpdated': lastUpdated,
+        'readme': readme,
+        'detectedTypes': detectedTypes,
+        'popularity': popularity,
+      };
+}
+
+class PackageQuery {
+  final String text;
+  final String type;
+  final int offset;
+  final int limit;
+
+  PackageQuery(this.text, {this.type, this.offset, this.limit});
+
+  factory PackageQuery.fromServiceQueryParameters(Map<String, String> params) {
+    final String text = params['q'];
+    String type = params['type'];
+    if (type != null && type.isEmpty) type = null;
+    int offset = int.parse(params['offset'] ?? '0', onError: (_) => 0);
+    int limit =
+        int.parse(params['limit'] ?? '0', onError: (_) => defaultSearchLimit);
+
+    offset = min(maxSearchResults - minSearchLimit, offset);
+    offset = max(0, offset);
+    limit = max(minSearchLimit, limit);
+
+    return new PackageQuery(text, type: type, offset: offset, limit: limit);
+  }
+
+  Map<String, String> toServiceQueryParameters() => {
+        'q': text,
+        'type': type,
+        'offset': offset?.toString(),
+        'limit': limit?.toString(),
+      };
+}
+
+class PackageSearchResult {
+  final String lastUpdated;
+  final int totalCount;
+  final List<PackageScore> packages;
+
+  PackageSearchResult({this.lastUpdated, this.totalCount, this.packages});
+
+  PackageSearchResult.notReady()
+      : lastUpdated = null,
+        totalCount = 0,
+        packages = [];
+
+  factory PackageSearchResult.fromJson(Map json) => new PackageSearchResult(
+        lastUpdated: json['lastUpdated'],
+        totalCount: json['totalCount'],
+        packages: (json['packages'] ?? [])
+            .map((Map m) => new PackageScore.fromJson(m))
+            .toList(),
+      );
+
+  /// Whether the search service has already updated its index after a startup.
+  bool get isLegit => lastUpdated != null;
+
+  Map toJson() => {
+        'lastUpdated': lastUpdated,
+        'totalCount': totalCount,
+        'packages': packages,
+      };
+}
+
+class PackageScore {
```

Comment:
A little nit: if/when we were to introduce a package-name-typeahead feature, and serve it through the search API, what would be the name of that?
Similar question for code search, but I strongly suspect that it can be part of the `PackageSearchResult`.